docs/design/event-streams.md

```diff
@@ -0,0 +1,246 @@
+# Kotlin Event Streaming Design
+
+* **Type**: Design
+* **Author(s)**: Aaron Todd
+
+# Abstract
+
+This document covers the client interfaces to be generated for requests/responses with event stream payloads 
+(`@streaming` Smithy trait applied to a `union` shape). See the [event stream spec](https://awslabs.github.io/smithy/1.0/spec/core/stream-traits.html#event-streams).
+
+Reference the additional documents listed in the Appendix for surrounding context on Smithy.
+
+
+# Design
+
+## Client Interface
+
+### Model
+
+We will use the examples from the event stream spec to explore the client interfaces to be generated:
+
+
+**Event stream on an input**
+
+```
+namespace smithy.example
+
+operation PublishMessages {
+    input: PublishMessagesInput
+}
+
+@input
+structure PublishMessagesInput {
+    room: String,
+    messages: PublishEvents,
+}
+
+@streaming
+union PublishEvents {
+    message: Message,
+    leave: LeaveEvent,
+}
+
+structure Message {
+    message: String,
+}
+
+structure LeaveEvent {}
+```
+
+**Event stream on an output**
+
+```
+namespace smithy.example
+
+operation SubscribeToMovements {
+    input: SubscribeToMovementsInput,
+    output: SubscribeToMovementsOutput
+}
+
+@input
+structure SubscribeToMovementsInput {}
+
+@output
+structure SubscribeToMovementsOutput {
+    movements: MovementEvents,
+}
+
+@streaming
+union MovementEvents {
+    up: Movement,
+    down: Movement,
+    left: Movement,
+    right: Movement,
+    throttlingError: ThrottlingError
+}
+
+structure Movement {
+    velocity: Float,
+}
+
+/// An example error emitted when the client is throttled
+/// and should terminate the event stream.
+@error("client")
+@retryable(throttling: true)
+structure ThrottlingError {}
+```
+
+
+### Event Stream Type Representation
+
+The members of an operation input or output that target a stream will be represented using by an asynchronous [Flow](https://kotlinlang.org/docs/reference/coroutines/flow.html) 
+from the `kotlinx-coroutines-core` library. `Flow` is a natural fit for representing asynchronous streams. 
+
+`Flow` was chosen for pagination and already in use as part of our public API contract. Any alternative to this would require a custom but similar type that doesn't play well with
+the rest of the coroutine ecosystem. There is also prior art for representing streaming requests and responses, see [gRPC Kotlin](https://github.com/grpc/grpc-kotlin).
```

Comment:
**Question**: "Prior art"?

Reply:
Are you asking what it is or the usage of it here?

Prior art just refers to the novelty of an idea/invention, often used in patent law when referring to the patentability of an idea.

In this case I'm pointing out that this _is not_ a novel idea and that others have used `Flow` to represent asynchronous streaming operations in a similar IDL based client/server implementation (gRPC). This serves as additional grounds/data that this approach has been proven out and has merit.

Reply:
Ah, got it. I hadn't heard this phrase before and guess it was an autocorrect/misspelling. We can leave it as-is.

Reply:
BTW, [rsocket-kotlin](https://github.com/rsocket/rsocket-kotlin) is also using `Flow` to represent streaming requests/responses over network.